### DISCUSSION TITLE
Testing a USB MIDI controller is working

### DISCUSSION THREAD

**Discussion by jshbrntt** [FundamentalFrequency/.github#36] [Q&A] [answered]
I currently have a Raspberry Pi Model B 4GB setup with Raspberry Pi OS 64 bit (Bullseye).

I have downloaded and can run without issue the the `aarch64` exectuables for the `LMN-3-DAW` and `LMN-3-Emulator` on the Pi.

Now I am attempting to use an existing USB MIDI controller (M-Audio Oxygen 8 v1) to control the DAW.

I can rebind the controls to the MIDI input the DAW expects but nothing is happening in the DAW (and there is no additional MIDI Input in the settings to select, other than the loopback interface used by the Emulator).

I've downloaded this project [ReceiveMIDI](https://github.com/gbevin/ReceiveMIDI) and am attempting to build it for `aarch64` so I can use it to debug if the USB MIDI controller is being registered at all by the Pi.

However, it requires Projucer which also needs to be built for `aarch64`.

I noticed in your repositories you have a project called [`juce-aarch64-linux-cross-compiler`](https://github.com/FundamentalFrequency/juce-aarch64-linux-cross-compiler).

Any chance of adding a bit of documentation on how to utilize this toolchain and `Dockerfile` to cross-compile JUCE for `aarch64`?

Also if you know of a easier way to debug a USB MIDI controller is being registered and working correctly with a Pi and sending the expected MIDI input.

**Comment by sramey** [comment]
Hi @joshua-barnett 

I'm super new to all of this so I'll share what I had came across when trying to do the same thing in Windows (So I could debug in Visual Studio and see values that were populating)

In MidiCommandManager.cpp around the switch statement, it is expecting certain values to come in after it notices that the message.isController()

```
case ENCODER_4_BUTTON:
            if (auto listener = dynamic_cast<Listener *>(focusedComponent)) {
                if (message.getControllerValue() == 127)
                    listener->encoder4ButtonPressed();

                if (message.getControllerValue() == 0)
                    listener->encoder4ButtonReleased();
            }
```

If the midi message was value ("static constexpr int ENCODER_4_BUTTON = 23;"), 23, then it checks for a value of 127 to say that it was pressed.

No idea if this helps you at all, that is as far as I got today.

**Comment by stonepreston** [comment, marked as answer]
The thing I would start with is running things on a amd64 linux installation first, and seeing if that picks it up. If you don't have that available, you could try going the VM route. If you dont want to do that, then the easiest thing would be forget about cross compilation as its a bit complex, and settle for compiling JUCE and its extras from source on the Pi. 4 GB should be enough RAM to compile JUCE from source. 

You can use the [BUILD_EXTRAS](https://github.com/juce-framework/JUCE/blob/2b16c1b94c90d0db3072f6dc9da481a9484d0435/CMakeLists.txt#L103) flag when generating the CMake build directory to build the JUCE extras. Part of these extras is something called the audio plugin host. You can use this host to test out midi receive by using one of the built in plugins which is called a MIDI listener (or something to that effect). [These docs](https://docs.juce.com/master/tutorial_create_projucer_basic_plugin.html#tutorial_create_projucer_basic_plugin_debugging) provide a basic look at using the plugin host. Its pretty straightforward.

**Comment by jshbrntt** [comment]
@stonepreston thanks for the advice, I compiled and ran it with the following commands.

```
$ cd JUCE
$ cmake . -B cmake-build -DJUCE_BUILD_EXAMPLES=ON -DJUCE_BUILD_EXTRAS=ON
$ cmake --build cmake-build --target AudioPluginHost
$ ./cmake-build/extras/AudioPluginHost/AudioPluginHost_artefacts/AudioPluginHost
```

I assume MIDI Logger is what I'm looking for?

However, when I select it nothing in the UI changes and there are no logs in the terminal (event when pressing on the Oxygen or the Emulator).

![2022-06-30-235757_800x480_scrot](https://user-images.githubusercontent.com/1833542/176792237-c6f0f8b7-7448-4e91-9f89-dfe3c7778267.png)

I'm guessing it logs there?

![2022-07-01-000020_800x480_scrot](https://user-images.githubusercontent.com/1833542/176792289-ac5d7466-dbad-4e3c-860d-cb7e5790eb32.png)

**Reply by stonepreston** [reply]
You will have to connect the midi ins and outs to the ports on the logger plugin. Then you can double click the plug-in and it should open a new window with the logs displayed. You might have to use the settings menu to select your midi controller (probably via the options menu up at the top of the plug-in host)

**Reply by jshbrntt** [reply]
Thanks I got it, the small screen size was obscuring things, I was able to delete all plugins and right click the empty space to add the MIDI Logger attached to the MIDI Input. Then double-click MIDI Logger to get the message log window.
![2022-07-01-002955_1299x994_scrot](https://user-images.githubusercontent.com/1833542/176795155-c212e867-aefa-45d0-bdf4-f75a140991ce.png)

However, it appears my MIDI USB controller is being identified as a potential input as the audio devices settings does not show an alternative input other than the emulator.

I guess I'll dive down the rabbit hole of trying to find Linux drivers for this outdated MIDI controller.

Thanks anyway.
